src/Common/src/System/Security/Cryptography/RsaPaddingProcessor.cs

```diff
@@ -0,0 +1,557 @@
+        {
+            _hashAlgorithmName = hashAlgorithmName;
+            _hLen = hLen;
+        }
+
+        internal static int BytesRequiredForBitCount(int keySizeInBits)
+        {
+            return (keySizeInBits + 7) / 8;
+        }
+
+        internal int HashLength => _hLen;
+
+        internal static RsaPaddingProcessor OpenProcessor(HashAlgorithmName hashAlgorithmName)
+        {
+            return s_lookup.GetOrAdd(
+                hashAlgorithmName,
+                alg =>
+                {
+                    using (IncrementalHash hasher = IncrementalHash.CreateHash(hashAlgorithmName))
+                    {
+                        // SHA-2-512 is the biggest we expect
+                        Span<byte> stackDest = stackalloc byte[512 / 8];
+
+                        if (hasher.TryGetHashAndReset(stackDest, out int bytesWritten))
+                        {
+                            return new RsaPaddingProcessor(hashAlgorithmName, bytesWritten);
+                        }
+
+                        byte[] big = hasher.GetHashAndReset();
+                        return new RsaPaddingProcessor(hashAlgorithmName, big.Length);
+                    }
+                });
+        }
+
+        internal static void PadPkcs1Encryption(
+            ReadOnlySpan<byte> source,
+            Span<byte> destination)
+        {
+            // https://tools.ietf.org/html/rfc3447#section-7.2.1
+
+            int mLen = source.Length;
+            int k = destination.Length;
+
+            // 1. If mLen > k - 11, fail
+            if (mLen > k - 11)
+            {
+                throw new CryptographicException(SR.Cryptography_KeyTooSmall);
+            }
+
+            // 2(b). EM is composed of 00 02 [PS] 00 [M]
+            Span<byte> mInEM = destination.Slice(destination.Length - source.Length);
+            Span<byte> ps = destination.Slice(2, destination.Length - source.Length - 3);
+            destination[0] = 0;
+            destination[1] = 2;
+            destination[ps.Length + 2] = 0;
+
+            // 2(a). Fill PS with random data from a CSPRNG, but no zero-values.
+            FillNonZeroBytes(ps);
+
+            source.CopyTo(mInEM);
+        }
+
+        internal void PadOaep(
+            ReadOnlySpan<byte> source,
+            Span<byte> destination)
+        {
+            // https://tools.ietf.org/html/rfc3447#section-7.1.2
+
+            byte[] dbMask = null;
+            Span<byte> dbMaskSpan = Span<byte>.Empty;
+
+            try
+            {
+                // Since the biggest known _hLen is 512/8 (64) and destination.Length is 0 or more,
+                // this shouldn't underflow without something having severely gone wrong.
+                int maxInput = checked(destination.Length - _hLen - _hLen - 2);
+
+                // 1(a) does not apply, we do not allow custom label values.
+
+                // 1(b)
+                if (source.Length > maxInput)
```

Comment:
I'm not certain I understand what's happening here for 1B. What we appear to be doing here is based on the assumption that `mLen` is correct, so the caller knows the destination length? 1C I get.

```
// Source = C
// Destination = M
int mLen = destination.Length;
int min_k = m_Len + 2*_hLen + 2; // from mLen <= k - 2hLen - 2

// 1(c)
// if (min_k < 2*_hLen + 2) throw cryptoException
// => if (destination.Length +2*_hLen + 2 < 2*_hLen + 2) 
// => if (destination.Length < 0)
// => always false because span.Length is always nonnegative.

// 1 (b)
// if (source.Length != k) throw cryptoException
// => if (source.Length < min_k) 
// => if (source.Length < destination.Length + 2*_hLen + 2)
// ... something here?
// => if (source.Length > destination.Length - 2*_hLen - 2)
```

I'm missing how to get to the assertion that your maxInput statement is correct.

When I flip source/destination the math makes sense, but then nothing I don't understand the parameters. That would mean that `source` holds the message and `destination` holds the ciphertext, so we would be doing the inverse of how the RFC defined the function.

```
// source = M
// destination = C 
int mLen = source.Length
int min_k = mLen + 2hlen +2 // from mLen <= k - 2hLen - 2

// 1(c)
// if (min_k < 2*_hLen + 2) throw cryptoException
// => if (source.Length +2*_hLen + 2 < 2*_hLen + 2) 
// => if (source.Length < 0)
// => always false because span.Length is always nonnegative.

// 1 (b)
// if (destination.Length != k) throw cryptoexception
// => if (destination.Length < min_k)
// => if (destination.Length < source.Length + 2hlen + 2)
// => if (source.Length > destination.Length - 2hlen - 2
```

Is that the case, or am I misunderstanding something fundamental here?

Reply:
lol, I was comparing this to the argument validation to the 7.1.2 section that you mentioned at the top. No wonder it didn't link up. Serves me right for trying to understand it line-by-line and not looking ahead :)